### PR TITLE
Raise at runtime if anything other than a keyword list is interpolated

### PR DESCRIPTION
**Pull request by parkerl**
Looking for some feedback. Thanks

- Issue #1089

### CODE REVIEW

File: lib/ecto/query/builder.ex [hunk 1]
```diff
@@ -65,10 +65,11 @@ defmodule Ecto.Query.Builder do
     {{:{}, [], [:fragment, [], [escaped]]}, params}
   end
 
-  def escape({:fragment, _, [{:^, _, _} = expr]}, _type, params, vars, env) do
-    {escaped, params} = escape(expr, :any, params, vars, env)
-
-    {{:{}, [], [:fragment, [], [escaped]]}, params}
+  def escape({:fragment, _, [{:^, _, [var]} = expr]}, _type, params, vars, env) do
```
Comment (lexmag) [inline lib/ecto/query/builder.ex]: There are several "unused variable" warnings.

File: lib/ecto/query/builder.ex [hunk 2]
```diff
@@ -218,6 +219,12 @@ defmodule Ecto.Query.Builder do
     error! "`#{Macro.to_string(other)}` is not a valid query expression"
   end
 
+  def runtime_fragment!([{_,_}|_] = kw) when is_list(kw), do: kw
+  def runtime_fragment!(expr) do
```
Comment (lexmag) [inline lib/ecto/query/builder.ex]: There is "unused variable" warning.

File: lib/ecto/query/builder.ex [hunk 3]
```diff
@@ -218,6 +219,12 @@ defmodule Ecto.Query.Builder do
...
+    raise ArgumentError, "Only a keyword list may be interpolated as the first " <>
```
Comment (lexmag) [inline lib/ecto/query/builder.ex]: Exception message should be non-capitalized and with no trailing punctuation.

File: test/ecto/query_test.exs [hunk 4]
```diff
@@ -266,4 +266,13 @@ defmodule Ecto.QueryTest do
     assert excluded_query.preloads == base.preloads
     assert excluded_query.assocs == base.assocs
   end
+
+  test "fragment/1 raises  at runtime when interpolation is not a keyword list" do
+    assert_raise ArgumentError, "Only a keyword list may be interpolated as the first " <>
+                                "argument to `fragment/1` with the `^` operator.",
+    fn ->
```
Comment (lexmag) [inline test/ecto/query_test.exs]: It'd be better to keep the `fn ->` on the previous line, as for now it's harder to follows `assert_raise` definition.
Maybe even better to assign message to variable before an assertion.

**Comment by parkerl** [comment]
Thanks @lexmag, I cleaned that up. What do you think of the overall implementation? Right track?

File: lib/ecto/query/builder.ex [hunk 5]
```diff
@@ -218,6 +219,12 @@ defmodule Ecto.Query.Builder do
     error! "`#{Macro.to_string(other)}` is not a valid query expression"
   end
 
+  def runtime_fragment!([{_,_}|_] = kw) when is_list(kw), do: kw
+  def runtime_fragment!(expr) do
+    raise ArgumentError, "to prevent sql injection, only a keyword list may be interpolated " <>
```
Comment (lexmag) [inline lib/ecto/query/builder.ex]: I think it should rather be `error!/1` call instead of plain `raise/2`.

File: lib/ecto/query/builder.ex [hunk 6]
```diff
@@ -218,6 +219,12 @@ defmodule Ecto.Query.Builder do
     error! "`#{Macro.to_string(other)}` is not a valid query expression"
   end
 
+  def runtime_fragment!([{_,_}|_] = kw) when is_list(kw), do: kw
```
Comment (lexmag) [inline lib/ecto/query/builder.ex]: Shall we consider empty list here as input?
We could use `Keyword.keyword?/1` function, don't think we expect really long lists there.

**Comment by lexmag** [comment]
@parkerl :+1:, looks like a right direction.

File: lib/ecto/query/builder.ex [hunk 7]
```diff
@@ -218,6 +219,12 @@ defmodule Ecto.Query.Builder do
...
+  def runtime_fragment!(expr) do
+    raise ArgumentError, "to prevent sql injection, only a keyword list may be interpolated " <>
+                         "as the first argument to `fragment/1` with the `^` operator; `#{expr}` was given"
```
Comment (josevalim) [inline lib/ecto/query/builder.ex]: "as the first argument to `fragment/1` with the `^` operator, got: `#{expr}`"

File: lib/ecto/query/builder.ex (hunk 5 above)
Comment (parkerl) [inline lib/ecto/query/builder.ex]: @lexmag I didn't use `error!/1` as it raises an `Ecto.Query.CompileError` which doesn't seem correct in this case. No?

Comment (josevalim) [inline lib/ecto/query/builder.ex]: That's what we use in other places, so I think we are fine: https://github.com/parkerl/ecto/blob/fragment_interpolation_validation/lib/ecto/query/builder.ex#L434

Comment (lexmag) [inline lib/ecto/query/builder.ex]: That't what I thought initially – `ArgumentError` should be in runtime, but [there is what made me think other way about](https://github.com/elixir-lang/ecto/blob/da82346505b6182ba62b7d560df30599dbe81e4c/lib/ecto/query/builder/update.ex#L154-L160).
Then we should fix those calls instead (not related to the current PR). :smile:

**Comment by parkerl** [comment]
What else do I need to do on this one? I think I've handled all the issues above. Thanks.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**